### PR TITLE
Add partitions ring support to Distributor.QueryStream()

### PR DESCRIPTION
**Pull request by pracucci**
#### What this PR does

In this PR I'm adding partitions ring support to `Distributor.QueryStream()`. Other Distributor's query functions don't use partitions ring yet and their support will be introduced in follow up PRs.

Notes:
- Most of work has been done by @dimitarvdimitrov , mentioned as co-author in the commit.
- I've introduced a new config option `-querier.prefer-availability-zone`. The config is hidden in the doc because currently used only by experimental ingest storage.

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/distributor/query_ingest_storage_test.go [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	ingester_client "github.com/grafana/mimir/pkg/ingester/client"
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	util_math "github.com/grafana/mimir/pkg/util/math"
+)
+
+func TestDistributor_QueryStream_ShouldSupportIngestStorage(t *testing.T) {
```
Comment (dimitarvdimitrov) [inline pkg/distributor/query_ingest_storage_test.go:26]: do you want to add test cases where ingesters have different states in the ring (LEAVING, ACTIVE)? 




```go
	// ringStates explicitly sets the state of each ingester in the ring.
	// If unset, instances will be in ring.ACTIVE state.
	ringStates []ring.InstanceState
```

I had these on the partitions-ring branch, but I see you omitted them in this PR

**Review by dimitarvdimitrov — COMMENTED** [review]
leaving a partial review, will continue reviewing tomorrow

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -1521,18 +1552,39 @@ func (d *Distributor) queryQuorumConfig(ctx context.Context, replicationSet ring
...
+		}
+
+		return zones
 	}
 }
 
 // LabelValuesForLabelName returns all of the label values that are associated with a given label name.
 func (d *Distributor) LabelValuesForLabelName(ctx context.Context, from, to model.Time, labelName model.LabelName, matchers ...*labels.Matcher) ([]string, error) {
-	replicationSet, err := d.GetIngesters(ctx)
+	//nolint:staticcheck
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/distributor/distributor.go:1586]: Yes, exactly.

[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/distributor/distributor.go:1586]: The usage of deprecated functions will disappear as soon as I complete this work (over the course of few more PRs).

File: pkg/distributor/query_ingest_storage_test.go (hunk 1 above)
Comment (pracucci) [reply pkg/distributor/query_ingest_storage_test.go:27]: Thanks for spotting it. It wasn't intentional. I did remove all test cases querying both partitions and "classic" ring (which is not supported in this PR) and I unintentionally missed tests on ingesters state. Re-added in this commit: https://github.com/grafana/mimir/pull/7388/commits/a1d4757dbaa0aa52ddc9ea4f82370a865df261de

File: pkg/distributor/query.go [hunk 3]
```diff
@@ -126,6 +131,49 @@ func (d *Distributor) GetIngesters(ctx context.Context) (ring.ReplicationSet, er
 	return d.ingestersRing.GetReplicationSetForOperation(readNoExtend)
 }
 
+// getIngesterReplicationSetForQuery returns a list of ring.ReplicationSet, containing ingester instances,
+// that must be queried for a read operation.
+//
+// If multiple ring.ReplicationSe are returned, each must be queried separately, and results merged.
```
Comment (pstibrany) [inline pkg/distributor/query.go:137]:
```suggestion
// If multiple ring.ReplicationSets are returned, each must be queried separately, and results merged.
```

File: pkg/distributor/query_ingest_storage_test.go [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+				},
+			},
+			preferZone:               "zone-a",
+			minimizeIngesterRequests: true,
+			matchers:                 []*labels.Matcher{selectAllSeriesMatcher},
+			expectedResponse:         expectedResponse(0, 0, false),
+			expectedQueriedIngesters: 5 /* zone-a */ + 2, /* the two failed fall back to zone-b */
+			expectedErr:              errFail,
+		},
+		"should succeed if there unhealthy ingesters in every zone but unhealthy ingesters own different partitions": {
```
Comment (pstibrany) [inline pkg/distributor/query_ingest_storage_test.go:246]: This is very nice property.

**Review by pstibrany — APPROVED** [review]
Very nice set of tests. Good job @dimitarvdimitrov and @pracucci!

File: pkg/distributor/query_ingest_storage_test.go (hunk 4 above)
Comment (pracucci) [reply pkg/distributor/query_ingest_storage_test.go:246]: The coolest property of the new system ;)

**Comment by pstibrany** [comment]
> > Very nice set of tests. Good job @dimitarvdimitrov
> 
> All done by Dimitar!

I thought so, but wasn't sure :)

**Comment by pracucci** [comment]
_Going to merge it since I don't think that what we do in this PR is particularly contentious. I there will be post merge comments, I will quickly address them in a follow up PR._

File: pkg/distributor/query_ingest_storage_test.go [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+			preferZone:               "zone-a",
+			minimizeIngesterRequests: true,
+			matchers:                 []*labels.Matcher{selectAllSeriesMatcher},
+			expectedErr:              ring.ErrTooManyUnhealthyInstances,
+		},
+		"should fail if all the ingesters owning a partition are in JOINING state": {
+			ingesterStateByZone: map[string]ingesterZoneState{
+				"zone-a": {numIngesters: 5, happyIngesters: 5, ringStates: []ring.InstanceState{ring.JOINING, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE}},
+				"zone-b": {numIngesters: 5, happyIngesters: 5, ringStates: []ring.InstanceState{ring.JOINING, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE, ring.ACTIVE}},
+			},
```
Comment (dimitarvdimitrov) [inline pkg/distributor/query_ingest_storage_test.go:413]: you can also add a case where different ingesters in the two zones are LEAVING (or JOINING) but the query still succeeds

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, left an idea for one more test case, but happy to do it in a follow-up PR if you don't want to wait for CI on this one again

File: pkg/distributor/query_ingest_storage_test.go (hunk 5 above)
Comment (pracucci) [reply pkg/distributor/query_ingest_storage_test.go:413]: Good idea, done in https://github.com/grafana/mimir/pull/7388/commits/19650e3de24bd7bfc6489de00c0ee5c55eee3972

**Pull request merged**